pkg/endpoint: wait for security identity on restore

aanm opened the pull request:
If the KVStore connectivity is not reliable during the endpoint restore
process Cilium can end up with an endpoint in a 'restoring' state in
case the ep's security identity resolution fails.
Adding a controller will make sure Cilium will retry to get an identity
for that endpoint until the endpoint is removed or the connectivity
with the allocator is successful.

Signed-off-by: André Martins <andre@cilium.io>

```release-note
avoid having endpoints in 'restoring' state in case the connectivity with the KVStore is not reliable
```

aanm committed:
pkg/endpoint: wait for security identity on restore

If the KVStore connectivity is not reliable during the endpoint restore
process Cilium can end up with an endpoint in a 'restoring' state in
case the ep's security identity resolution fails.
Adding a controller will make sure Cilium will retry to get an identity
for that endpoint until the endpoint is removed or the connectivity
with the allocator is successful.

Signed-off-by: André Martins <andre@cilium.io>

joestringer commented on pkg/endpoint/restore.go:
```diff
@@ -222,13 +223,35 @@ func (e *Endpoint) restoreIdentity() error {
+	controllerName := fmt.Sprintf("restoring-ep-identity (%v)", e.ID)
+	var (
+		identity *identity.Identity
+		gotIdty  = make(chan struct{})
```
nit:

```suggestion
		allocatedIdentity  = make(chan struct{})
```

joestringer replied:
(Mostly nitpicking on the `Idty` which is not that much shorter but presumably means `Identity` so we should be able to just use `Identity`)

aanm committed:
pkg/endpoint: wait for global security identities on restore

If the KVStore connectivity is not reliable during the endpoint restore
process Cilium can end up with an endpoint in a 'restoring' state in
case the global security identities sync would fail or time out.
Adding a controller will make sure Cilium will wait until the global
security identities are synced or until the endpoint is removed before
restoring the endpoint.

Signed-off-by: André Martins <andre@cilium.io>

joestringer commented on pkg/endpoint/restore.go:
```diff
@@ -222,28 +223,72 @@ func (e *Endpoint) restoreIdentity() error {
+		e.UpdateController(controllerName,
+			controller.ControllerParams{
+				DoFunc: func(ctx context.Context) (err error) {
+					identityCtx, cancel := context.WithTimeout(context.Background(), option.Config.KVstoreConnectivityTimeout)
```
```suggestion
					identityCtx, cancel := context.WithTimeout(ctx, option.Config.KVstoreConnectivityTimeout)
```

joestringer commented on pkg/endpoint/restore.go:
```diff
@@ -222,28 +223,72 @@ func (e *Endpoint) restoreIdentity() error {
-		defer cancel()
+		// Getting the initial global identities while we are restoring should
+		// block the restoring of the endpoint.
+		// If the endpoint is removed, this controller will cancel the allocator
+		// WaitForInitialGlobalIdentities function.
```
How is this step implemented? I see the exit case below to continue if `aliveCtx` is cancelled, but I don't see cleanup of the new controller.

aanm replied:
@joestringer all controllers in `e.controllers` are cleaned up when the endpoint is disconnected. Is part of the disconnect function.

joestringer reviewed (changes requested):
Looks mostly fine to me, I inspected the callpath up the stack from here and it seems like this is run in a goroutine and completion is notified back through a channel to the main goroutine via `restoreComplete` but this does not block overall Cilium operation, only some subsequent cleanup steps.

For sanity it probably makes sense to add some documentation higher up in those paths to declare how long we expect that channel could block for so we don't end up introducing dependencies there in future that could cause the whole agent to block up completely when kvstore connectivity is completely blocked (or at least to make sure it integrates properly with the initial kvstore connect timeout of 15m).

My main feedback request here is I don't understand in the case of failure (specifically endpoint deletion) how these controllers are cleaned up.
